source/traefik_proxy.go

```diff
@@ -0,0 +1,903 @@
+	for _, hostname := range hostnameList {
+		endpoints = append(endpoints, endpointsForHostname(hostname, targets, ttl, providerSpecific, setIdentifier)...)
+	}
+
+	for _, route := range ingressRoute.Spec.Routes {
+		match := route.Match
+
+		for _, hostEntry := range traefikHostExtractor.FindAllString(match, -1) {
+			for _, host := range traefikValueProcessor.FindAllString(hostEntry, -1) {
+				host = strings.TrimPrefix(host, "`")
+				host = strings.TrimSuffix(host, "`")
+
+				// Checking for host = * is required, as HostSNI(`*`) can be set
+				// in the case of TLS passthrough
+				if host != "*" && host != "" {
+					endpoints = append(endpoints, endpointsForHostname(host, targets, ttl, providerSpecific, setIdentifier)...)
+				}
+			}
+		}
+	}
+
+	return endpoints, nil
+}
+
+// endpointsFromIngressRouteUDP extracts the endpoints from a IngressRouteUDP object
+func (ts *traefikSource) endpointsFromIngressRouteUDP(ingressRoute *IngressRouteUDP, targets endpoint.Targets) ([]*endpoint.Endpoint, error) {
+	var endpoints []*endpoint.Endpoint
+
+	providerSpecific, setIdentifier := getProviderSpecificAnnotations(ingressRoute.Annotations)
+
+	ttl, err := getTTLFromAnnotations(ingressRoute.Annotations)
+	if err != nil {
+		return nil, err
+	}
+
+	hostnameList := getHostnamesFromAnnotations(ingressRoute.Annotations)
+	for _, hostname := range hostnameList {
+		endpoints = append(endpoints, endpointsForHostname(hostname, targets, ttl, providerSpecific, setIdentifier)...)
+	}
+
+	return endpoints, nil
+}
+
+func (ts *traefikSource) AddEventHandler(ctx context.Context, handler func()) {
+	// Right now there is no way to remove event handler from informer, see:
+	// https://github.com/kubernetes/kubernetes/issues/79610
+	log.Debug("Adding event handler for IngressRoute")
+	ts.ingressRouteInformer.Informer().AddEventHandler(eventHandlerFunc(handler))
+	log.Debug("Adding event handler for IngressRouteTCP")
+	ts.ingressRouteTcpInformer.Informer().AddEventHandler(eventHandlerFunc(handler))
+	log.Debug("Adding event handler for IngressRouteUDP")
+	ts.ingressRouteUdpInformer.Informer().AddEventHandler(eventHandlerFunc(handler))
+}
+
+// newTraefikUnstructuredConverter returns a new unstructuredConverter initialized
+func newTraefikUnstructuredConverter() (*unstructuredConverter, error) {
+	uc := &unstructuredConverter{
+		scheme: runtime.NewScheme(),
+	}
+
+	// Add the core types we need
+	uc.scheme.AddKnownTypes(ingressrouteGVR.GroupVersion(), &IngressRoute{}, &IngressRouteList{})
+	uc.scheme.AddKnownTypes(ingressrouteTCPGVR.GroupVersion(), &IngressRouteTCP{}, &IngressRouteTCPList{})
+	uc.scheme.AddKnownTypes(ingressrouteUDPGVR.GroupVersion(), &IngressRouteUDP{}, &IngressRouteUDPList{})
+	if err := scheme.AddToScheme(uc.scheme); err != nil {
+		return nil, err
+	}
+
+	return uc, nil
+}
+
+// Basic redefinition of Traefik 2's CRD: https://github.com/traefik/traefik/tree/v2.8.7/pkg/provider/kubernetes/crd/traefik/v1alpha1
```

Comment:
How do you plan to keep those in sync? As much as I am not a fan of having a direct dependency on trafik, I think that importing traefik directly might make it easier to keep the dependencies up to date as the process would resolve to a simple `go get`. WDYT?

Reply:
In the initial implementation, I had a direct dependency on traefik but removed it later down the line ([in this commit](https://github.com/kubernetes-sigs/external-dns/pull/3055/commits/fd8a4f33ed872d80a73f1f165343b307aa4f47d5)).

The Traefik maintainers were not a big fan of the hard dependency either.
I was recommended to go for a redefinition approach as is done with [gloo](https://github.com/kubernetes-sigs/external-dns/blob/master/source/gloo_proxy.go#L47).
This way, external-dns can stay compatible with multiple Traefik CRD versions without immediately dropping support when performing a `go get`, in case Traefik would perform significant changes on their side.

Also, as only a subset of the fields is relevant for external-dns, staying in sync should be fine unless breaking changes to the CRDs happen, i.e., the routes and match paths in the CRD change.
Anyhow, I'm open to both keeping it redefined and importing it directly.

Reply:
how is the progress on this conversation? Do you guys need any help with this?
This PR is the only blocker we have for switching back to Kubernetes CRD on Traefik.